crawler/twitter.py

```diff
@@ -0,0 +1,183 @@
+
+        for user in results:
+            twitterCredentials.append(
+                {"userName": user["userName"],
+                "accessToken": user["accessToken"],
+                "accessTokenSecret": user["accessTokenSecret"]}
+            )
+
+        return twitterCredentials
+
+    def __get_twitter_users_queue(self):
+        usersQueue = []
+        results = self.db.run("MATCH (n:UsersQueue) RETURN n.id as id")
+        for user in results:
+            usersQueue.append({"id": user["id"]})
+
+        return usersQueue
+
+    def __validate_twitter_user(self, twitterHelper, userId):
+        api = TwitterAPI(
+            self.token,
+            self.secret,
+            twitterHelper["accessToken"],
+            twitterHelper["accessTokenSecret"])
+
+        response = api.request(self.USERS_SHOW, {"user_id": userId})
+        parsedResponse = json.loads(response.text)
+        isValid = False
+        if "status" in parsedResponse:
+            isValid = self.__is_language_valid(parsedResponse["status"]["lang"])
+        elif "errors" in parsedResponse:
+            isValid = self.__check_twitter_error_code(
+                parsedResponse["errors"][0]["code"]
+            )
+
+        return isValid
+
+    def __is_language_valid(self, language):
+        return {
+            self.GERMAN: True
+        }.get(language, False)
+
+    def __check_twitter_error_code(self, code):
+        return {
+            32: "Cannot authenticate",
+            50: "User not found",
+            63: "Suspended user",
+            88: "Switch helper",
+            89: "Switch helper",
+            131: "Internal error"
+        }.get(code, "Unknown error")
+
+    def __delete_twitter_user_from_queue(self, id):
+        query = "MATCH (n:UsersQueue {id: '" + id + "'}) "
+        query += "DETACH DELETE n"
+        response = self.db.run(query)
+        print("Twitter user with id: ", id, " deleted from users queue.")
+
+    def __create_users_queue_batch_query(self, key, followerId, followersCount):
+        if key == 0 and followersCount == 0:
+            key = "singleFollower"
+
+        query = "(:UsersQueue {id: '" + followerId + "'})"
+
+        return {
+            0: "CREATE " + query + ", ",
+            followersCount: query,
+            "singleFollower": "CREATE " + query
+        }.get(key, query + ", ")
+
+    def __create_user_followers_query(self, userId, followerId):
+        query = "MERGE (u:TwitterUser{id:'" + userId + "'}) "
+        query += "MERGE (f:TwitterUser{id:'" + followerId + "'}) "
+        query += "MERGE (u)<-[:FOLLOWS]-(f)"
+
+        return query
+
+    def __create_user_followers_batch_query(
+        self, key, userId, followerId, followersCount):
+
+        if key == 0 and followersCount == 0:
+            key = "singleFollower"
+
+        query = "MERGE (u"+str(key)+":TwitterUser{id:'" + userId + "'}) "
+        query += "MERGE (f"+str(key)+":TwitterUser{id:'" + followerId + "'}) "
```

Comment:
I don't know if it should be here, but all the user info/metadata (screen_name, created_at, followers_count, ...) that we want to store in the nodes should enter the nodes as a property dictionary, e.g.  "MERGE (f"+str(key)+":TwitterUser{uid:'" + followerId + "', screen_name: '" + followerScreenName + "'}) " Do we include that in the crawler?

Reply:
I think we can add this matadata to the user, but not its followers. To fetch this data you need to make a call to twitter api and some users have thousands of followers, this would quickly exhaust all the helpers. I still think its a good idea though, if we can collect more twitter token donations from people we can definitely do this.

Reply:
I think thats what bruno wanted to say. We add it just for the crawled user, because every user is crawled, we would finally have all the user information in the database.
I actually dont like the users queue approach.
Have a look at the issue i created here: #32 

This approach works without any users queue, because this object needs a lot of memory, also its not possible to run crawlers with a usersqueue in parallel.
A central instance collecting the oldest nodes or uncrawled nodes could also be the solution for users not in our database. 
If someone generates a tracemap and the author or some users are not in our database, they could be prioritized by this central instance.
For example we could have a flask route /crawler/get_new_users which returns 10 new user ids for the crawler. in that way we wouldnt have to worry about redundant crawlers and can be quite flexible because of different modules which are easily exchangeable.